Merge src/Native/build.sh into top level build.sh

janhenke opened the pull request:
If no argument is passed, defaults to build both the managed and the
native code. The src/Native/build.sh is kept to not break CI, it simply
calls the top level build.sh to call the native part. Closes #3155 

I tested on Linux.

janhenke commented:
@mmitche The native PR tests fail due to the way how src/Native/build.sh is called. I left the file there, which just calls the top level build.sh now. Shall I fix this problem or can we agree on updating the PR check configuration to simply use the top level build.sh? In theory we also do not need separate jobs for the managed and native parts any more.

jasonwilliams200OK commented:
:+1: 
Perhaps we can remove `src/Native/build.sh` and send a PR for @mmitche fixing https://github.com/dotnet/dotnet-ci/search?utf8=%E2%9C%93&q=src%2FNative%2Fbuild.sh&type=Code.

jasonwilliams200OK commented:
@dotnet-bot, test this please.

janhenke commented:
Okay I have updated the PR and dropped the other build.sh completely. I will add a PR to the dotnet/dotnet-ci repository to update the job configurations accordingly.

janhenke commented:
Okay this PR has to be merged together with dotnet/dotnet-ci#59, then everything should work again and we have proper "one builds it all" support.

nguerrera commented on build.sh:
```diff
@@ -1,96 +1,353 @@
+        ;;
+
+    *)
+        echo "Unsupported OS $OSName detected, configuring as if for Linux"
```
I know this is how it was, but we should make this an error and remove OpenBSD and NetBSD until there's an actual effort to port to those.

janhenke replied:
I would not make it an hard error. But I am okay with removing OpenBSD and NetBSD for now.

nguerrera commented:
I would do these three changes sequentially:

1. This change + src/Native/build.sh stub that works when invoked from any directory (*)
2. dotnet/dotnet-ci#59
3. Delete src/Native/build.sh stub

That way, there's no race to get this and dotnet/dotnet-ci#59 in together.

(*) I believe all that was needed was the same __scriptDir approach that was already there instead of using ../.. from the current directory.

mmitche commented:
@jasonmalinowski @janhenke Unfortunately a PR sent to dotnet-ci won't directly affect the CI.  It's mostly currently used a backup location.  There is stuff in the works to shift jobs to being generated by the Job DSL plugin which will enable this kind of workflow, but we're not there yet.  For now I'd go with Nick's approach

janhenke commented:
I have updated the PR. The src/Native/build.sh is back as a stub calling the top level build.sh. This time this should work from the top level as well. So we can forward with the three step plan as mentioned by Nick above.

jasonwilliams200OK commented:
:+1: for @nguerrera's approach. Thanks @janhenke for the prompt update.
@mmitche, is it so that in future we will have access to Job DSL setting in repository and we can include CI changes within the related PR? That would be nice. This is exactly how other GitHub repositories do which make use of TravisCI+AppVeyor stack. :+1:

mmitche commented:
@jasonwilliams200OK Yeah that's the plan.  I'm thinking that most of the configuration will still go into dotnet-ci, so that a lot of common configuration parameters can be pulled out.

nguerrera commented on build.sh:
```diff
@@ -1,96 +1,345 @@
+    echo "Usage: $0 [managed] [native] [BuildArch] [BuildType] [clean] [verbose] [clangx.y]"
+    echo "manged - optional argument to build the managed code"
+    echo "native - optional argument to build the native code"
+    echo "The following arguments affect native builds only:"
```
Just a thought for future, but Clean, Verbose, and BuildType could (and probably should) be made to work with managed too.

janhenke replied:
Correct, but I tried to keep the delta small with this PR. We can always extend the functionality from this point on.

nguerrera commented:
LGTM. Can someone else take a look as well? 

cc @ellismg

ellismg commented:
LGTM.  Thanks @janhenke.  @mmitche Are you okay with this being merged now?

stephentoub commented:
@mmitche, can this be merged?

mmitche commented:
Yes.
